File: src/algorithms/graph/traversal/bfs/bfs.test.ts
import { describe, it, expect } from "vitest";

import { breadthFirstSearch } from "./sources/bfs.ts?fn";

type AdjacencyList = Record<string, string[]>;

describe("breadthFirstSearch", () => {
  it("traverses a linear graph in order", () => {
    const adjacencyList: AdjacencyList = {
      A: ["B"],
      B: ["C"],
      C: ["D"],
      D: [],
    };
    expect(breadthFirstSearch(adjacencyList, "A")).toEqual(["A", "B", "C", "D"]);
  });

  it("traverses a tree graph level by level", () => {
    const adjacencyList: AdjacencyList = {
      A: ["B", "C"],
      B: ["D", "E"],
      C: ["F"],
      D: [],
      E: [],
      F: [],
    };
    const result = breadthFirstSearch(adjacencyList, "A");
    expect(result).toEqual(["A", "B", "C", "D", "E", "F"]);
  });

  it("handles a disconnected graph by visiting only reachable nodes", () => {
    const adjacencyList: AdjacencyList = {
      A: ["B"],
      B: [],
      C: ["D"],
      D: [],
    };
    const result = breadthFirstSearch(adjacencyList, "A");
    expect(result).toEqual(["A", "B"]);
    expect(result).not.toContain("C");
    expect(result).not.toContain("D");
  });

  it("handles a single node graph", () => {
    const adjacencyList: AdjacencyList = {
      A: [],
    };
    expect(breadthFirstSearch(adjacencyList, "A")).toEqual(["A"]);
  });

  it("does not visit the same node twice in a cyclic graph", () => {
    const adjacencyList: AdjacencyList = {
      A: ["B"],
      B: ["C"],
      C: ["A"],
    };
    const result = breadthFirstSearch(adjacencyList, "A");
    expect(result).toEqual(["A", "B", "C"]);
  });

  it("visits neighbors in the order they appear in the adjacency list", () => {
    const adjacencyList: AdjacencyList = {
      A: ["C", "B"],
      B: [],
      C: [],
    };
    const result = breadthFirstSearch(adjacencyList, "A");
    expect(result).toEqual(["A", "C", "B"]);
  });

  it("handles a node with no neighbors listed in the adjacency list", () => {
    const adjacencyList: AdjacencyList = {
      A: ["B"],
    };
    const result = breadthFirstSearch(adjacencyList, "A");
    expect(result).toEqual(["A", "B"]);
  });

  it("traverses a fully connected graph", () => {
    const adjacencyList: AdjacencyList = {
      A: ["B", "C", "D"],
      B: ["A", "C", "D"],
      C: ["A", "B", "D"],
      D: ["A", "B", "C"],
    };
    const result = breadthFirstSearch(adjacencyList, "A");
    expect(result).toHaveLength(4);
    expect(result[0]).toBe("A");
    expect(new Set(result)).toEqual(new Set(["A", "B", "C", "D"]));
  });
});

File: src/algorithms/graph/traversal/bfs/__tests__/BFS_test.cpp
#include "../sources/BFS.cpp"
#include <algorithm>
#include <cassert>
#include <iostream>

int main() {
    // Test 1: linear graph
    {
        unordered_map<string, vector<string>> adj = {{"A",{"B"}},{"B",{"C"}},{"C",{"D"}},{"D",{}}};
        assert((BFS::breadthFirstSearch(adj, "A") == vector<string>{"A","B","C","D"}));
    }

    // Test 2: tree level by level
    {
        unordered_map<string, vector<string>> adj = {
            {"A",{"B","C"}},{"B",{"D","E"}},{"C",{"F"}},{"D",{}},{"E",{}},{"F",{}}
        };
        assert((BFS::breadthFirstSearch(adj, "A") == vector<string>{"A","B","C","D","E","F"}));
    }

    // Test 3: disconnected graph
    {
        unordered_map<string, vector<string>> adj = {{"A",{"B"}},{"B",{}},{"C",{"D"}},{"D",{}}};
        auto result = BFS::breadthFirstSearch(adj, "A");
        assert((result == vector<string>{"A","B"}));
        assert(find(result.begin(), result.end(), "C") == result.end());
    }

    // Test 4: single node
    {
        unordered_map<string, vector<string>> adj = {{"A",{}}};
        assert((BFS::breadthFirstSearch(adj, "A") == vector<string>{"A"}));
    }

    // Test 5: cyclic graph — no duplicates
    {
        unordered_map<string, vector<string>> adj = {{"A",{"B"}},{"B",{"C"}},{"C",{"A"}}};
        assert((BFS::breadthFirstSearch(adj, "A") == vector<string>{"A","B","C"}));
    }

    // Test 6: neighbor order preserved
    {
        unordered_map<string, vector<string>> adj = {{"A",{"C","B"}},{"B",{}},{"C",{}}};
        assert((BFS::breadthFirstSearch(adj, "A") == vector<string>{"A","C","B"}));
    }

    // Test 7: node missing from adjacency list
    {
        unordered_map<string, vector<string>> adj = {{"A",{"B"}}};
        auto result = BFS::breadthFirstSearch(adj, "A");
        assert((result == vector<string>{"A","B"}));
    }

    // Test 8: fully connected graph
    {
        unordered_map<string, vector<string>> adj = {
            {"A",{"B","C","D"}},{"B",{"A","C","D"}},
            {"C",{"A","B","D"}},{"D",{"A","B","C"}}
        };
        auto result = BFS::breadthFirstSearch(adj, "A");
        assert(result.size() == 4);
        assert(result[0] == "A");
    }

    cout << "All tests passed!" << endl;
    return 0;
}
